pkg/simple_cache/lib/simple_cache.dart

```diff
@@ -0,0 +1,219 @@
+import 'dart:async';
+import 'dart:convert';
+
+import 'package:identity_codec/identity_codec.dart';
+import 'package:logging/logging.dart';
+import 'cache_provider.dart';
+import 'src/providers/inmemory.dart';
+import 'src/providers/redis.dart' as redis;
+
+final _logger = Logger('simple_cache');
+
+/// Cache for objects of type [T], wrapping a [CacheProvider] to provide a
+/// high-level interface.
+///
+/// Cache entries are accessed using the indexing operator `[]`, this returns a
+/// [Entry<T>] wrapper that can be used to get/set data cached at given key.
+///
+/// **Example**
+/// ```dart
+/// final Cache<List<int>> cache = Cache.inMemoryCache(4096);
+///
+/// // Write data to cache
+/// await cache['cached-zeros'].set([0, 0, 0, 0]);
+///
+/// // Read data from cache
+/// var r = await cache['cached-zeros'].get();
+/// expect(r, equals([0, 0, 0, 0]));
+/// ```
+///
+/// A [Cache] can be _fused_ with a [Codec] using [withCodec] to get a cache
+/// that stores a different kind of objects. It is also possible to create
+/// a chuild cache using [withPrefix], such that all entries in the child
+/// cache have a given prefix.
+abstract class Cache<T> {
+  /// Get [Entry] wrapping data cached at [key].
+  Entry<T> operator [](String key);
+
+  /// Get a [Cache] wrapping this cache with given [prefix].
+  Cache<T> withPrefix(String prefix);
+
+  /// Get a [Cache] wrapping this cache while logging operations to [logger].
+  ///
+  /// If [ignoreErrors] is `false` (default), errors are logged as warnings
+  /// with [logger.warning()] and propagated to the caller.
+  ///
+  /// If [ignoreErrors] is `true`, errors are logged are logged as errors with
+  /// [logger.severe()] and ignored, instead calls to `Entry.get` will return
+  /// `null`.
+  ///
+  /// Reading and writing to the cache is always logged with [logger.finest()].
+  ///
+  /// If [withLogger] is called again the previous [logger] is overwritten.
+  Cache<T> withLogger(Logger logger, {bool ignoreErrors = false});
+
+  /// Get a [Cache] wrapping this cache by encoding objects of type [S] as [T]
+  /// using the given [codec].
+  Cache<S> withCodec<S>(Codec<S, T> codec);
+
+  /// Get a [Cache] wrapping this cache with given [ttl] as default for all
+  /// entries being set using [Entry.set].
+  ///
+  /// This only specifies a different default [ttl], to be used when [Entry.set]
+  /// is called without a [ttl] parameter.
+  Cache<T> withTTL(Duration ttl);
+
+  /// Create a [Cache] wrapping a [CacheProvider].
+  factory Cache(CacheProvider<T> provider) {
+    return _Cache<T, T>(provider, '', _logger, false, IdentityCodec());
+  }
+
+  /// Create an in-memory [CacheProvider] holding a maximum of [maxSize] cache
+  /// entries.
+  static CacheProvider<List<int>> inMemoryCacheProvider(int maxSize) {
```

Comment:
This part is a bit confusing: `Cache.something` is used to create a `CacheProvider`, and also has a `Cache(provider)` constructor. I'd probably expect only the later, and maybe another factory constructor for the in-memory cache, and the redis-based implementation should be separate from this one.

Reply:
If we decide that all `CacheProvider` implementations have a `close()` method, then this becomes rather simple.

Reply:
I correct myself: we need this to expose access to `close()`